Review HTTP/2 GOAWAY handling

sbordet opened the pull request.

sbordet committed:
Issue 5310 - Review HTTP/2 GOAWAY handling.

Moved stream "close" event after returning from listener methods.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue 5310 - Review HTTP/2 GOAWAY handling.

Refactored reset code in a single place: HTTP2Session.reset().

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue 5310 - Review HTTP/2 GOAWAY handling.

Refactored push code in a single place: HTTP2Session.push().

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue 5310 - Review HTTP/2 GOAWAY handling.

Code cleanups: removed unnecessary final modifiers and fixed code warnings.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue 5310 - Review HTTP/2 GOAWAY handling.

Reimplemented close/idle_timeout/stop/onGoAway/input_shutdown following more closely the specification.

In particular, the semantic of sending a GOAWAY is now to:
* stop creation of new both local and remote streams
* record the last processed stream
* continue processing streams that are pending

This means that a GOAWAY is "graceful" in the sense that it allows for streams to be completed by applications.

The semantic of stop() and idle timeout is harsher: for pending streams a RST_STREAM is sent to the other peer and they are failed locally.

Added support for GOAWAY with 2^31-1 lastStreamId.

Added support for a peer to send and receive multiple GOAWAY frames.

Reviewed the stream creation/destruction mechanism so that when the last stream completes after a GOAWAY, proper actions can be run to tear down the connection.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue 5310 - Review HTTP/2 GOAWAY handling.

Updates after initial review.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue 5310 - Review HTTP/2 GOAWAY handling.

Fixed logic in onShutdown(): in case of abort the closing action must be
run as callback, so it executes at the end of all the other callbacks.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue 5310 - Review HTTP/2 GOAWAY handling.

Fixed javadocs and TODOs.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented on jetty-http2/http2-client/src/main/java/org/eclipse/jetty/http2/client/HTTP2ClientConnectionFactory.java:
```diff
@@ -92,20 +92,6 @@ private HTTP2ClientConnection(HTTP2Client client, ByteBufferPool byteBufferPool,
             this.listener = listener;
         }
 
-        @Override
-        public long getMessagesIn()
-        {
-            HTTP2ClientSession session = (HTTP2ClientSession)getSession();
-            return session.getStreamsOpened();
-        }
-
-        @Override
-        public long getMessagesOut()
-        {
-            HTTP2ClientSession session = (HTTP2ClientSession)getSession();
-            return session.getStreamsClosed();
-        }
-
```
why were these removed?

gregw commented on jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Flusher.java:
```diff
@@ -365,7 +365,7 @@ protected void onCompleteFailure(Throwable x)
         // flusher, we need to close the connection.
         if (closed == null)
-            session.abort(x);
+            session.onWriteFailure(x);
```
comment is not correct as you are not directly closing.
How is onWriteFailure different to abort?

gregw reviewed (commented):
might get you to walk me through the new tests tomorrow...

sbordet committed:
Fixes #5633 - Allow to configure HttpClient request authority.

Fixed initial session recv window update: it was wrong if the initial value was less than the default value (65535).

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixes #5633 - Allow to configure HttpClient request authority.

Updated after review.
Added more tests.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue 5310 - Review HTTP/2 GOAWAY handling.

Updates after review.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw reviewed (approved):
LGTM

gregw commented:
But don't merge until 9.4.35 is promoted

lorban reviewed (approved):
LGTM
